components/ledger/pkg/storage/sqlstorage/ledger/migrates/0-init-schema/any.go

```diff
@@ -0,0 +1,514 @@
+package initschema
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+
+	"github.com/formancehq/ledger/pkg/core"
+	"github.com/formancehq/ledger/pkg/ledger/aggregator"
+	"github.com/formancehq/ledger/pkg/storage"
+	ledgerstore "github.com/formancehq/ledger/pkg/storage/sqlstorage/ledger"
+	"github.com/formancehq/ledger/pkg/storage/sqlstorage/migrations"
+	"github.com/formancehq/ledger/pkg/storage/sqlstorage/schema"
+	"github.com/formancehq/stack/libs/go-libs/metadata"
+	"github.com/lib/pq"
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"github.com/uptrace/bun"
+)
+
+func init() {
+	migrations.RegisterGoMigration(UpgradeLogs)
+}
+
+const (
+	LogTableName = "log"
+
+	LogsMigrationBatchSizeFlag = "logs-migration-batch-size"
+	OldSchemaRenameSuffixFlag  = "old-schema-rename-prefix"
+)
+
+var (
+	batchSize             uint64 = 10000
+	oldSchemaRenameSuffix        = "_save_v2_0_0"
+)
+
+func InitMigrationConfigCLIFlags(flags *flag.FlagSet) {
+	flags.Uint64(LogsMigrationBatchSizeFlag, 10000, "Batch size for logs migration")
+	flags.String(OldSchemaRenameSuffixFlag, "save_v2_0_0", "Name of the old schema (to be renamed)")
+
+	b := viper.GetUint64(LogsMigrationBatchSizeFlag)
+	if b != 0 {
+		batchSize = b
+	}
+
+	suffix := viper.GetString(OldSchemaRenameSuffixFlag)
+	if suffix != "" {
+		oldSchemaRenameSuffix = suffix
+	}
+}
+
+type Log struct {
+	bun.BaseModel `bun:"log,alias:log"`
+
+	ID   uint64          `bun:"id,unique,type:bigint"`
+	Type string          `bun:"type,type:varchar"`
+	Hash string          `bun:"hash,type:varchar"`
+	Date core.Time       `bun:"date,type:timestamptz"`
+	Data json.RawMessage `bun:"data,type:jsonb"`
+}
+
+func isLogTableExisting(
+	ctx context.Context,
+	schema schema.Schema,
+	sqlTx *schema.Tx,
+) (bool, error) {
+	row := sqlTx.QueryRowContext(ctx, fmt.Sprintf(`
+	SELECT EXISTS (
+		SELECT FROM
+			pg_tables
+		WHERE
+			schemaname = '%s' AND
+			tablename  = 'log'
+		)
+	`, schema.Name()))
+	if row.Err() != nil {
+		return false, errors.Wrap(row.Err(), "checking if log table exists")
+	}
+
+	var exists bool
+	if err := row.Scan(&exists); err != nil {
+		return false, errors.Wrap(err, "scanning if log table exists")
+	}
+
+	return exists, nil
+}
+
+func readLogsRange(
+	ctx context.Context,
+	schema schema.Schema,
+	sqlTx *schema.Tx,
+	idMin, idMax uint64,
+) ([]Log, error) {
+	rawLogs := make([]Log, 0)
+	sb := schema.
+		NewSelect(LogTableName).
+		Where("id >= ?", idMin).
+		Where("id < ?", idMax).
+		Model((*Log)(nil))
+
+	rows, err := sqlTx.QueryContext(ctx, sb.String())
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return rawLogs, nil
+		}
+
+		return nil, errors.Wrap(err, "selecting logs")
+	}
+
```

Comment:
Maybe add a defer to close rows if the loop fail.

Reply:
Done thanks !